pkg/odo/cli/utils/convert.go

```diff
@@ -0,0 +1,416 @@
+	if err != nil {
+		return nil, "", err
+	}
+	imageStream, err := client.GetImageStream(imageNS, imageName, imageTag)
+	if err != nil {
+		return nil, "", err
+	}
+
+	imageStreamImage, err := client.GetImageStreamImage(imageStream, imageTag)
+	if err != nil {
+		return nil, "", err
+	}
+
+	imageforDevfile := imageStream.Spec.Tags[0].From.Name
+
+	return imageStreamImage, imageforDevfile, nil
+}
+
+// setDevfileCommandsForS2I sets command in devfile.yaml from s2i data.
+func setDevfileCommandsForS2I(d data.DevfileData) {
+
+	buildCommand := common.DevfileCommand{
+		Exec: &common.Exec{
+			Id:          buildCommandID,
+			Component:   containerName,
+			CommandLine: buildCommandS2i,
+			Group: &common.Group{
+				Kind:      common.BuildCommandGroupType,
+				IsDefault: true,
+			},
+		},
+	}
+
+	runCommand := common.DevfileCommand{
+		Exec: &common.Exec{
+			Id:          runCommandID,
+			Component:   containerName,
+			CommandLine: runCommandS2i,
+			Group: &common.Group{
+				Kind:      common.RunCommandGroupType,
+				IsDefault: true,
+			},
+		},
+	}
+
+	// Ignoring error as we are writing new file
+	_ = d.AddCommands([]common.DevfileCommand{buildCommand, runCommand})
+
+}
+
+// setDevfileComponentsForS2I sets the devfile.yaml components field from s2i data.
+func setDevfileComponentsForS2I(d data.DevfileData, s2iImage string, localConfig *config.LocalConfigInfo, s2iEnv config.EnvVarList) error {
+
+	maxMemory := localConfig.GetMaxMemory()
+	volumes := localConfig.GetStorage()
+	// list of ports taken from builder image and set into local config
+	ports := localConfig.GetPorts()
+
+	var endpoints []common.Endpoint
+	var envs []common.Env
+	var volumeMounts []common.VolumeMount
+	var components []common.DevfileComponent
+
+	// convert s2i storage to devfile volumes
+	for _, vol := range volumes {
+		volume := common.Volume{
+			Name: vol.Name,
+			Size: vol.Size,
+		}
+		components = append(components, common.DevfileComponent{Volume: &volume})
+
+		volumeMount := common.VolumeMount{
+			Name: vol.Name,
+			Path: vol.Path,
+		}
+
+		volumeMounts = append(volumeMounts, volumeMount)
+	}
+
+	// Add s2i specific env variable in devfile
+	for _, env := range s2iEnv {
+		env := common.Env{
+			Name:  env.Name,
+			Value: env.Value,
+		}
+		envs = append(envs, env)
+	}
+	env := common.Env{
+		Name:  envS2iConvertedDevfile,
+		Value: "true",
+	}
+	envs = append(envs, env)
+
+	// convert s2i ports to devfile endpoints
+	for _, port := range ports {
+
+		port := strings.Split(port, "/")
+		// from s2i config.yaml port is of the form 8080/TCP
+		// Ignoring TCP and Udp values here
+		portInt, err := strconv.ParseInt(port[0], 10, 32)
+		if err != nil {
+			return errors.Wrapf(err, "Unable to convert port %s from config.yaml to devfile.yaml", port)
+		}
+
+		endpoint := common.Endpoint{
+			Name:       fmt.Sprintf("port-%s", port[0]),
+			TargetPort: int32(portInt),
+			Configuration: &common.Configuration{
```

Comment:
I think configuration was removed from the schema recently https://github.com/devfile/kubernetes-api/blob/master/schemas/devfile.json#L439